nimble/host/services/audio/pacs/src/ble_audio_svc_pacs.c

```diff
@@ -0,0 +1,428 @@
+        } , {
+            .uuid = BLE_UUID16_DECLARE(
+                BLE_SVC_AUDIO_PACS_CHR_UUID16_SINK_AUDIO_LOCATIONS),
+            .access_cb = ble_svc_audio_pacs_access,
+            .flags = BLE_GATT_CHR_F_READ_ENC,
+        } , {
+            .uuid = BLE_UUID16_DECLARE(
+                BLE_SVC_AUDIO_PACS_CHR_UUID16_SOURCE_PAC),
+            .access_cb = ble_svc_audio_pacs_access,
+            .flags = BLE_GATT_CHR_F_READ_ENC
+        } , {
+            .uuid = BLE_UUID16_DECLARE(
+                BLE_SVC_AUDIO_PACS_CHR_UUID16_SOURCE_AUDIO_LOCATIONS),
+            .access_cb = ble_svc_audio_pacs_access,
+            .flags = BLE_GATT_CHR_F_READ_ENC,
+        }, {
+            .uuid = BLE_UUID16_DECLARE(
+                BLE_SVC_AUDIO_PACS_CHR_UUID16_AVAILABLE_AUDIO_CONTEXTS),
+            .access_cb = ble_svc_audio_pacs_access,
+            .flags = BLE_GATT_CHR_F_READ_ENC | BLE_GATT_CHR_F_NOTIFY,
+        }, {
+            .uuid = BLE_UUID16_DECLARE(
+                BLE_SVC_AUDIO_PACS_CHR_UUID16_SUPPORTED_AUDIO_CONTEXTS),
+            .access_cb = ble_svc_audio_pacs_access,
+            .flags = BLE_GATT_CHR_F_READ_ENC,
+        }, {
+            0, /* No more characteristics in this service */
+        }
+        }
+    },
+    {
+        0, /* No more services. */
+    },
+};
+
+static int
+codec_record_to_pacs_entry(struct ble_audio_codec_record *record, void *arg)
+{
+    struct pac_read_cb_arg *cb_arg = (struct pac_read_cb_arg *)arg;
+    int rc;
+
+    if (!(record->flags & cb_arg->flags)) {
+        return 0;
+    }
+
+    rc = os_mbuf_append(cb_arg->om, &record->codec_id,
+                        sizeof(struct ble_audio_codec_id));
+    if (rc) {
+        return BLE_ATT_ERR_INSUFFICIENT_RES;
+    }
+
+    rc = os_mbuf_append(cb_arg->om, &record->codec_spec_caps_len,
+                   sizeof(uint8_t));
+    if (rc) {
+        return BLE_ATT_ERR_INSUFFICIENT_RES;
+    }
+
+    rc = os_mbuf_append(cb_arg->om, &record->codec_spec_caps, record->codec_spec_caps_len);
+    if (rc) {
+        return BLE_ATT_ERR_INSUFFICIENT_RES;
+    }
+
+    rc = os_mbuf_append(cb_arg->om, &record->metadata_len,
+                   sizeof(uint8_t));
+    if (rc) {
+        return BLE_ATT_ERR_INSUFFICIENT_RES;
+    }
+
+    rc = os_mbuf_append(cb_arg->om, &record->metadata, record->metadata_len);
+    if (rc) {
+        return BLE_ATT_ERR_INSUFFICIENT_RES;
+    }
+
+    cb_arg->pac_count++;
+
+    return 0;
+}
+
+static int
+ble_svc_audio_pacs_sink_pac_read_access(struct ble_gatt_access_ctxt *ctxt)
+{
+    struct pac_read_cb_arg cb_arg = {
+        .flags = BLE_AUDIO_CODEC_FLAG_SINK,
+        .om = ctxt->om,
+        .pac_count = 0
+    };
+    int rc;
+    uint8_t *pac_count;
+
+    pac_count = os_mbuf_extend(ctxt->om, sizeof(uint8_t));
+    rc = ble_audio_codec_foreach(codec_record_to_pacs_entry, (void *)&cb_arg);
+
+    *pac_count = cb_arg.pac_count;
+
+    return rc == 0 ? 0 : BLE_ATT_ERR_INSUFFICIENT_RES;
+}
```

Comment:
Hmm, that does not look right to me, as I think you should send as many data as you could fit into the buffer, so that the remote could perform Long Value Read (incl. offset > 0).

Reply:
This callback fills whole buffer with full value, and offset is taken into considerations layer up, so only part of what was read here is actually sent. Full call chain is like this:
ble_att_svr_rx_read_blob -> ble_att_svr_read_handle -> ble_att_svr_read -> entry->ha_cb (ble_gatts_chr_val_access) -> ble_gatts_val_access -> access_cb

`ble_gatts_val_access`  has code like this:
```c
        rc = access_cb(conn_handle, attr_handle, gatt_ctxt, cb_arg); \\ This is where our app (service) callback reads whole value
        if (rc == 0) {
            attr_len = OS_MBUF_PKTLEN(gatt_ctxt->om) - initial_len - offset;
            if (attr_len >= 0) {
                if (new_om) {
                    os_mbuf_appendfrom(*om, gatt_ctxt->om, offset, attr_len); \\ Here value from offset is copied into rsp buffer
                }
            } else {
                rc = BLE_ATT_ERR_INVALID_OFFSET;
            }
        }
```

Reply:
On this level all access OP codes are:
`BLE_GATT_ACCESS_OP_READ_CHR`
`BLE_GATT_ACCESS_OP_WRITE_CHR`
`BLE_GATT_ACCESS_OP_READ_DSC`
`BLE_GATT_ACCESS_OP_WRITE_DSC`
And we do not distinguish reads, writes etc